google/cloud/spanner/internal/logging_spanner_stub.cc

```diff
@@ -0,0 +1,240 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/internal/logging_spanner_stub.h"
+#include "google/cloud/internal/invoke_result.h"
+#include "google/cloud/log.h"
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+namespace {
+
+template <typename T>
+struct IsStatusOr : public std::false_type {};
+template <typename T>
+struct IsStatusOr<StatusOr<T>> : public std::true_type {};
+
+template <typename T>
+struct IsUniquePtr : public std::false_type {};
+template <typename T>
+struct IsUniquePtr<std::unique_ptr<T>> : public std::true_type {};
+
+template <
+    typename Functor, typename Request,
+    typename Result = google::cloud::internal::invoke_result_t<
+        Functor, grpc::ClientContext&, Request const&>,
+    typename std::enable_if<std::is_same<Result, google::cloud::Status>::value,
+                            int>::type = 0>
+Result LogHelper(Functor&& functor, grpc::ClientContext& context,
+                 Request const& request, char const* where) {
+  GCP_LOG(DEBUG) << where << "() << " << request.DebugString();
+  auto response = functor(context, request);
+  GCP_LOG(DEBUG) << where << "() >> status=" << response;
+  return response;
+}
+
+template <typename Functor, typename Request,
+          typename Result = google::cloud::internal::invoke_result_t<
+              Functor, grpc::ClientContext&, Request const&>,
+          typename std::enable_if<IsStatusOr<Result>::value, int>::type = 0>
+Result LogHelper(Functor&& functor, grpc::ClientContext& context,
+                 Request const& request, char const* where) {
+  GCP_LOG(DEBUG) << where << "() << " << request.DebugString();
+  auto response = functor(context, request);
+  if (!response) {
+    GCP_LOG(DEBUG) << where << "() >> status=" << response.status();
+  } else {
+    GCP_LOG(DEBUG) << where << "() >> response=" << response->DebugString();
+  }
+  return response;
+}
+
+template <typename Functor, typename Request,
+          typename Result = google::cloud::internal::invoke_result_t<
+              Functor, grpc::ClientContext&, Request const&>,
+          typename std::enable_if<IsUniquePtr<Result>::value, int>::type = 0>
+Result LogHelper(Functor&& functor, grpc::ClientContext& context,
+                 Request const& request, char const* where) {
+  GCP_LOG(DEBUG) << where << "() << " << request.DebugString();
+  return functor(context, request);
```

Comment:
Do you think it might still be useful to log whether the returned unique_ptr was nullptr or not?

Reply:
Don't know.  In every case these are `std::unique_ptr<grpc::ClientReader<T>>`, i.e., the result of a streaming (read) RPC.  As far as I can tell those are never null: gRPC indicates an error by returning a (non-null) reader, whose `Finish()` function tells you what the error was.

But it is easy to add, and may help in timing, and would be less surprising, so why not?

Reply:
I'll leave it to you. I didn't realize the limited cases where unique_ptr was used here, so it may not be that useful. I only noticed that this overload looked different from the others and was wondering why.

Again, your call if you think it's useful.

Reply:
... oh, and you added it.. OK. Cool.